### ISSUE TITLE
Call `HyperRouter.redirect` on initialization

### ISSUE THREAD

**Issue by strash** [sufftea/hyper_router#1]
Hi,
I love the api, but have some problems. Maybe I'm doing something wrong.

I have an Auth view and the shell in the root of the router. I've put first tab from the shell in the `initialRoute` and even if the user is not authorized I get this route. I suppose I could handle this situation using the `redirect` function, but it only fires when the user navigates to a different route.

### Example: 

``` dart
// ignore_for_file: use_build_context_synchronously

import 'package:flutter/material.dart';
import 'package:hive_flutter/hive_flutter.dart';
import 'package:hyper_router/hyper_router.dart';

// ------ nothing important here ------

Future<void> main() async {
  WidgetsFlutterBinding.ensureInitialized();

  await Hive.initFlutter();
  await Hive.openBox("auth");

  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp.router(
      routerConfig: router,
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
    );
  }
}

// --------- important part starts here ---------

// -> router

final router = HyperRouter(
  initialRoute: Tab1.routeName,
  redirect: (context, state) {
    // = IMPORTANT: =================================================
    // Redirect should be called on the router initialization.
    // Even if the user is not authorized we are getting Tab1 anyway.
    // ==============================================================
    final isAuthorized =
        Hive.box("auth").get("is_authorized", defaultValue: false);
    if (!isAuthorized) return Auth.routeName;
    return null;
  },
  routes: [
    // -> auth
    NamedRoute(
      name: Auth.routeName,
      screenBuilder: (_) => const Auth(),
    ),
    // -> tabs
    ShellRoute(
      shellBuilder: (context, controller, child) {
        return Scaffold(
          body: child,
          bottomNavigationBar: NavigationBar(
            selectedIndex: controller.tabIndex,
            onDestinationSelected: (idx) => controller.setTabIndex(idx),
            destinations: const [
              NavigationDestination(icon: Icon(Icons.home), label: "Tab 1"),
              NavigationDestination(icon: Icon(Icons.wc), label: "Tab 2"),
            ],
          ),
        );
      },
      tabs: [
        // -> tab 1
        NamedRoute(
          name: Tab1.routeName,
          screenBuilder: (_) => const Tab1(),
          children: [
            // -> subview
            ValueRoute<Tab1SubViewPayload>(
              screenBuilder: (_, value) => Tab1SubView(value),
            ),
          ],
        ),
        // -> tab 2
        NamedRoute(
          name: Tab2.routeName,
          screenBuilder: (_) => const Tab2(),
          children: [
            // -> subview
            ValueRoute<Tab2SubViewPayload>(
              screenBuilder: (_, value) => Tab2SubView(value),
            ),
          ],
        ),
      ],
    ),
  ],
);

// -> auth

class Auth extends StatefulWidget {
  static const routeName = RouteName("auth");

  const Auth({super.key});

  @override
  State<Auth> createState() => _AuthState();
}

class _AuthState extends State<Auth> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text("Authorization")),
      body: Center(
        child: ElevatedButton(
          onPressed: () async {
            await Hive.box("auth").put("is_authorized", true);
            if (mounted) context.hyper.navigate(Tab1.routeName);
          },
          child: const Text("Sign in"),
        ),
      ),
    );
  }
}

// -> tab 1

class Tab1 extends StatefulWidget {
  static const routeName = RouteName("tab1");

  const Tab1({super.key});

  @override
  State<Tab1> createState() => _Tab1State();
}

class _Tab1State extends State<Tab1> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text("Tab 1")),
      body: Center(
        child: ElevatedButton(
          onPressed: () =>
              context.hyper.navigate(Tab1SubViewPayload("Subview 1")),
          child: const Text("Open subview within tab 1"),
        ),
      ),
    );
  }
}

// -> tab 2

class Tab2 extends StatefulWidget {
  static const routeName = RouteName("tab2");

  const Tab2({super.key});

  @override
  State<Tab2> createState() => _Tab2State();
}

class _Tab2State extends State<Tab2> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text("Tab 2")),
      body: Center(
        child: ElevatedButton(
          onPressed: () =>
              context.hyper.navigate(Tab2SubViewPayload("Subview 2")),
          child: const Text("Open subview within tab 2"),
        ),
      ),
    );
  }
}

// ----- never mind. I figured this part right after I wrote this code :) -----
// -> subview

class Tab1SubViewPayload extends RouteValue {
  final String title;
  Tab1SubViewPayload(this.title);
}

class Tab2SubViewPayload extends RouteValue {
  final String title;
  Tab2SubViewPayload(this.title);
}

abstract class BaseSubView extends StatefulWidget {
  const BaseSubView({super.key});

  RouteValue get value;
  String get title;
}

class Tab1SubView extends BaseSubView {
  @override
  final Tab1SubViewPayload value;

  const Tab1SubView(this.value, {super.key});

  @override
  String get title => value.title;

  @override
  State<Tab1SubView> createState() => _SubViewState();
}

class Tab2SubView extends BaseSubView {
  @override
  final Tab2SubViewPayload value;

  const Tab2SubView(this.value, {super.key});

  @override
  String get title => value.title;

  @override
  State<Tab2SubView> createState() => _SubViewState();
}

class _SubViewState<T extends BaseSubView> extends State<T> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text(widget.title)),
      body: Center(
        child: ElevatedButton(
          onPressed: () => Navigator.of(context).pop(),
          child: const Text("Go back"),
        ),
      ),
    );
  }
}
```

**Comment by strash** [comment]
Apart from this the router works as expected. I tested it on my current medium sized project and compared to the go_router it's just so good! Thank you!

**Comment by sufftea** [comment]
Glad you liked the api:)

I haven't considered redirecting instantly, my bad. This should be easy to fix though.

For now you can check for authentication somewhere in initState and call navigate there.

Btw, I didn't cover this in the docs, but there are two ways you might want to do guarding: 
- Using the redirect callback, which only reacts to navigation
- Or by creating some sort of a state listener (e.g. a BlocListener) externally and navigating to auth when it's triggered, which makes the redirect reactive. Check out example/main.dart.

**Comment by sufftea** [comment]
Should be fixed on v0.0.3

**Comment by strash** [comment]
> Btw, I didn't cover this in the docs, but there are two ways you might want to do guarding:
> 
> * Using the redirect callback, which only reacts to navigation
> * Or by creating some sort of a state listener (e.g. a BlocListener) externally and navigating to auth when it's triggered, which makes the redirect reactive. Check out example/main.dart.

I saw the example, but I think instant redirect might cover the second way. It just feels right.
